### ISSUE TITLE
See Also lines right after usage lines?

### ISSUE THREAD

**Issue by adrianVmariano** [revarbat/BOSL2#1084]
It seems like the "See Also" lines are often after the arguments list, which means they are kind of lost, especially if there is a long description and lots of args.  Should they instead appear after the Usage, along with Topics?

**Comment by RAMilewski** [comment]
I think they should be at the end of the Description.  I'm also of the opinion that Description ought to be the topmost paragraph(s) for each function.

For newcomers just the name may not be enough to encourage them to scroll through lengthy (i.e. intimidating) Usage sections.

**Comment by adrianVmariano** [comment]
I wonder if you're confused.  The "Description" is sometimes 3 screens long.  The Usage is usually one or two lines except in a few cases.  It would be ridiculous to put the Description before the Usage.  The whole point of the Usage is to give a short quick summary of how to use the function.  It wouldn't do its job if it was buried after other stuff.  

Consider this case, for example:

https://github.com/revarbat/BOSL2/wiki/skin.scad#functionmodule-path_sweep

Isn't the See Also kinda lost after the Arguments section?  And wouldn't it be just as bad sandwiched between Arguments and Description?   So putting See Also at the top, before all the long potentially overwhelming stuff makes it easier to notice and easy to find if you want to look for it.

**Comment by revarbat** [comment]
The `See Also:` line feels logically (to me) like it should follow the description, but I agree it may get lost after the `Arguments:` tables.

**Comment by RAMilewski** [comment]
Sure, I'll take a crack at writing them.  I'll need to refresh my memory on doing a PR.

**Comment by RAMilewski** [comment]
> I would argue that module usage and function usage are pretty similar.

I agree, but in the interest of avoiding the **TL;DR** problem, I would suggest just describing the module action in the Synopsis for most entries.

**Comment by revarbat** [comment]
In many cases, a synopsis like "Constructs an XXX" covers both the module and function form, since the function form makes the VNF equivalent of the module's geometry.

**Comment by adrianVmariano** [comment]
That's basically what I was thinking.  So do you agree that this project makes sense and you'll have support for a Synopsis header?

**Comment by revarbat** [comment]
I like the idea of `Synopsis:` lines.  I can tweak docsgen to show the synopsis for functions and modules in the Alphabetic Index.  Another idea that has been bouncing around the back of my mind is representative images for geometry generating modules, so you can see a small torus image before the Usage lines, but after the Synopsis.

**Comment by adrianVmariano** [comment]
Ok.  So then RAMilewski is good to go to write Synopsis lines.

**Comment by adrianVmariano** [comment]
Sounds like you're in business.  Hold off on threading.scad and skin.scad, as I have changes for those files.  

If you do happen to notice errors or other deficiencies in the docs feel free to fix.

**Comment by RAMilewski** [comment]
I tried adding **DefineHeader(Text): Synopsis** to .openscad_docsgen_rc and the synopses now show, but not in the correct location.  

I also relocated See Also in the .scad file, but in most (but not all) cases its location is unchanged.

What controls the order of docsgen blocks in the .md output file?

**Comment by adrianVmariano** [comment]
I've never looked at docsgen, but Revar said that some of the headers have their order forced somehow, so it's not just the order they appear in the code.  Revar is working on updating docsgen so it should be ready soon.

**Comment by RAMilewski** [comment]
Ok.  Once we know the order I can rework the .scad file to match (just to preserve sanity).   I was thinking that it could be the Synopsis followed by See Also followed by Topics.   That puts all the things adjacent that someone might need to refer to in order to understand the module.    No big deal for anything in transforms.scad (except maybe frame_map() and apply()).  But more important as we get to more advanced modules.

**Comment by adrianVmariano** [comment]
If you group the one-line entries at the top and let docsgen produce the desired order that's probably fine.   

A look at the contents of "Attachable Masks" reveals that the section is misnamed.  It doesn't contain any masks.  It's contains mask attachers.  I'm not sure of the best name for that.  The actual masks are the 3d masks, I think.  And the mask attachers can be used to attach them.

**Comment by revarbat** [comment]
I've released openscad_docsgen v2.0.34 ( https://github.com/revarbat/openscad_docsgen/releases/tag/v2.0.34 ).
It supports the Synopsis: line, and fixes the sort order of blocks in Module/Function/Constant docs.

**Comment by revarbat** [comment]
I've added Synopsis lines to transforms.scad.

**Comment by RAMilewski** [comment]
Thanks for the openscad_docsgen update!

**Comment by adrianVmariano** [comment]
As you work through this you can also add Topics entries if it seems appropriate.

**Comment by adrianVmariano** [comment]
I suspect that there's a Topics index somewhere and that tutorials links wouldn't really make sense in the topics lists.  There may be some other way to add such links.  Perhaps the See Also entry could do it, though that probably requires a docsgen alteration.

**Comment by RAMilewski** [comment]
> A look at the contents of "Attachable Masks" reveals that the section is misnamed.

How about changing that to "Mask Attachment"?

**Comment by adrianVmariano** [comment]
Sounds reasonable to me.

**Comment by RAMilewski** [comment]
In shapes2d.scad both star() and supershape() take an atype parameter with two valid choices "hull" and "intersect".
However there is no explanation of the difference between the two and no relevant examples.

A search of the Attachments Tutorial also came up empty.   Did I miss something somewhere?

**Comment by adrianVmariano** [comment]
Make a new issue for this.   I think it's a big topic that needs work for a satisfactory solution.  "intersect" anchoring anchors to the intersection point of a ray in the anchor direction, and "hull" anchoring finds a point on the convex hull of the shape in the anchor direction.

**Comment by adrianVmariano** [comment]
I took a brief scan over your PR and you're definitely on the right track.  I have a few quibbles, but basically it's looking good.  

So quibbles: stroke can take a region.  (dashed stroke basically calls stroke, so they do the same thing except for the dashing.)  And turtle3d can return a list of transforms.  For star just say "star shaped region" and skip the "N points".  

It looks like you reverted some doc tweaks I made to prismoid, though not a huge deal.  (I tried to tighten the list of usage options and added some see also entries.)  

Over all, I wonder if you should say "Creates a Fooblat object or returns a path" or something to that effect.  That is, never mentioning that the function form returns a path may hide this behavior from users who are skimming the docs.  For the 3d objects it could say "or returns a VNF" when appropriate.

**Comment by RAMilewski** [comment]
Thanks!  That helps.  I'll make another pass through what I've done so far before I do another PR

**Comment by RAMilewski** [comment]
It looks like nothing in partitions.scad appears in the topics lists.

**Comment by adrianVmariano** [comment]
I don't really have anything to do with the topics.  Go ahead and add topics if it seems appropriate.

**Comment by revarbat** [comment]
Topics were another late addition to the documentation, and not everything has been given topics yet.

**Comment by RAMilewski** [comment]
I'll keep an eye open for that as I move forward.   I set the things in partitions.scad to Topics: Partitions, Masking

**Comment by revarbat** [comment]
Of note, a blank line after a See Also or Topics line will truncate the docs for the function/module right there, which is not what you want.

**Comment by RAMilewski** [comment]
OUCH!   Sorry!   I'll fix that going forward.

**Comment by RAMilewski** [comment]
OK... my testing shows it's just completely blank links.
Empty comment lines seem to work ok.

**Comment by revarbat** [comment]
Here's a checklist of files and whether or not they have Topics, See Alsos, and Synopses.  I'm starting it blank, and we can check off each file that is verified:

- [ ] affine.scad
- [ ] attachments.scad
- [ ] ball\_bearings.scad
- [ ] beziers.scad
- [ ] bottlecaps.scad
- [ ] builtins.scad
- [ ] color.scad
- [ ] comparisons.scad
- [ ] constants.scad
- [ ] coords.scad
- [ ] cubetruss.scad
- [ ] distributors.scad
- [ ] drawing.scad
- [ ] fnliterals.scad
- [ ] gears.scad
- [ ] geometry.scad
- [ ] hinges.scad
- [ ] joiners.scad
- [ ] linalg.scad
- [ ] linear\_bearings.scad
- [ ] lists.scad
- [ ] masks2d.scad
- [ ] masks3d.scad
- [ ] math.scad
- [ ] metric\_screws.scad
- [ ] modular\_hose.scad
- [ ] mutators.scad
- [ ] nema\_steppers.scad
- [ ] partitions.scad
- [ ] paths.scad
- [ ] polyhedra.scad
- [ ] regions.scad
- [ ] rounding.scad
- [ ] screw\_drive.scad
- [ ] screws.scad
- [ ] shapes2d.scad
- [ ] shapes3d.scad
- [ ] skin.scad
- [ ] sliders.scad
- [ ] strings.scad
- [ ] structs.scad
- [ ] threading.scad
- [ ] transforms.scad
- [ ] trigonometry.scad
- [ ] tripod\_mounts.scad
- [ ] turtle3d.scad
- [ ] utility.scad
- [ ] vectors.scad
- [ ] version.scad
- [ ] vnf.scad
- [ ] walls.scad
- [ ] wiring.scad

**Comment by revarbat** [comment]
I'm targeting updating files not in std.scad for the near future.

**Comment by RAMilewski** [comment]
As usual, I'm confused.  Does that mean you'd like be to leave things in std.scad to you, or start with those myself going forward.

**Comment by revarbat** [comment]
The latter. I plan to work on the docs of files not listed in std.scad. At least until we get nearer to complete.

**Comment by adrianVmariano** [comment]
Uh oh, someone did threading.scad, which has massive changes....

**Comment by adrianVmariano** [comment]
Maybe merging will be not to hard since I haven't updated the usage messages.

**Comment by RAMilewski** [comment]
How do Tutorials make it into the wiki?  I tried using openscad-docsgen on another project and everything seems to happen as expected except the tutorials.  Did I miss a command line argument somewhere?

**Comment by revarbat** [comment]
The openscad_docsgen package also installs the script `openscad-mdimggen` which you can run with no arguments to generate the tutorials.

**Comment by revarbat** [comment]
Okay, I've run out of std.scad files.  My next target will be vectors.scad, and I will move forward through the alphabet from there.

**Comment by adrianVmariano** [comment]
Is there a better way to convey the information other than "or returns a VNF" on the synopsis lines?  I didn't write that on the entries for rounding.scad, and to fit it in and keep the synopses to one line I have to cut out useful description.   It does seem useful, though, to have it clear which ones can return a VNF and which don't.   Could we have some shorthand symbol for return type in the synopsis?  `[GEOM] [VNF]`  The user could hover over it to get elaboration, e.g. "Returns geometry" or "Returns a VNF".  Maybe some even more useful mouseover text could be crafted.

**Comment by RAMilewski** [comment]
I like the idea of using rollover text.  I'll have to look up how to make that work in Github Markdown though.

**Comment by adrianVmariano** [comment]
Revar has done it elsewhere.  We'd need support for this to happen automatically anyway, or writing the synposis lines will be too painful.

**Comment by RAMilewski** [comment]
Maybe not too hard .  Stack Overflow says: 

This is a [hover text](## "your hover text") example.

**Comment by RAMilewski** [comment]
Should be able to just do a search and replace to fix "or returns a VNF."

**Comment by RAMilewski** [comment]
> We'd need support for this to happen automatically anyway, or writing the synposis lines will be too painful.

Why do you think we'd need support?

**Comment by adrianVmariano** [comment]
I think that you're being short-sighted and not thinking about the long term maintenance and big picture here.  If I have to explicitly write out a hover text or 2 every time I write a synopsis line, that's a PITA.   How do you ensure that the hover texts are all the same?  What if you need to change/update the hover text?

**Comment by revarbat** [comment]
I do already have hover-text `STD` FileFootnotes: tags in the wiki Sidebar.  They are not currently implemented in the cleanest way, however.  I am thinking of implementing a DefineFileFootnotes: configuration block that will let you do something like this in the config file, though:

```
DefineFileFootnotes:
    STD=Included in std.scad
    Parts=Can create parts.
    MAT=Support for transformation matrices.
```

Then in each file:

```
FileFootnotes: STD,MAT
```

Similarly, I am thinking of `DefineTags:` and `Tags:` for Constants/Function/Module documentation. The tags can be shown in the same places in the indices that we currently show Synopsis: line text.

I should note that we currently do have a number of Function&Modules that have a Topics entry for "VNF Generator".

**Comment by RAMilewski** [comment]
For repetitive hover texts like [VNF] you can just input a placeholder such as xyzzy and then do a search and replace to put in the standard hover text syntax everywhere it's used in a file.   If there's a way for docsgen to do the replacing then that's a win.  

Personally, I favor a hover text rather than a footnote link that takes the reader away from the module or function in question.   

FWIW the STD links in the sidebar  don't appear to point to actual content yet.   ...at least not in Firefox or Chrome.

**Comment by adrianVmariano** [comment]
If I write a new function next month and need to put 3 hover texts in its summary then I cannot use your search and replace scheme.  That's what I mean about short term thinking.   Yeah, if the library is done, never more to change, and we're writing the final documentation, then your approach of search and replace works.  But if it's a dynamic work, that's going to keep changing, and you need to be able to write and rewrite the tags in the future, then it does not.

**Comment by RAMilewski** [comment]
I'm all for letting docsgen do the fiddly bits!  So let's define what the mouse-over links look like before we write any more synopses.

[vnf], [geom], and [vnf || geom]  perhaps?

**Comment by adrianVmariano** [comment]
The compound idea [vnf || geom] is a mistake.  It seems like we need at least [vnf], [geom], [path], [rgn] (or [region]).   Not sure if that covers it?  I don't know how far we want to push this.  Like we could have things like [list] or [array] or [number], [boolean]....

**Comment by RAMilewski** [comment]
Not sure the compound idea is worthless. Since we can define each different mouseover only once at the beginning of the file, having some common compound mouseover links would simplify the synopses.
It's a little hard to show here because we're inside Github Markdown, but....

This code (minus the backslashes) near the top of the markdown:
`
  \[vnf]: ## "Used as a function it returns a vnf."

  \[geom]: ## "Used as a function it returns geometry."

  \[vnf||geom]: ## "Used as a function it can return a vnf or geometry."

  \[mat]: ## "Used as a function it returns a transformation matrix."
`

Works like this: 

  [vnf]: ## "Used as a function it returns a vnf."
  [geom]: ## "Used as a function it returns geometry."
  [vnf||geom]: ## "Used as a function it can return a vnf or geometry."
  [mat]: ## "Used as a function it returns a transformation matrix."

Synopsis: This module creates an elephant. [VNF][vnf] 

Synopsis: This module creates a tiger. [more info][vnf||geom]

**Comment by RAMilewski** [comment]
Hmmm... github got me again!   The Synopsis lines as written look like:

Synopsis: This module creates an elephant. [VNF][vnf] 

Synopsis: This module creates a tiger. [more info][vnf||geom]

**Comment by RAMilewski** [comment]
Complication:  None of this survives docsgen...   yet.

**Comment by adrianVmariano** [comment]
A function cannot return geometry.  So the mouseover for [geom] should be "Can be used as a module to return geometry" and then the mouseover for [vnf] would be "Can be used as a function to return a VNF".  The problem with compound entries is that then you have to worry about all possible combinations.  What if something can be geometry, or return a VNF or transformation matrix?  What if it returns a path or vnf or geometry?   I think each return should be separate.  

There is a question of whether there should be [geom2d] "can be used as a module to return 2d geometry" and [geom3d] analogously.  Or maybe just [2d] and [3d] would work for that?

**Comment by RAMilewski** [comment]
The synopsis describes the most common usage, with the mouseovers providing completeness.
For many entries the synopsis is describing the module. 
You know the library far better than I, which module or function has the largest number of different returns?

**Comment by adrianVmariano** [comment]
At the moment I can't think of anything with more than 3 return cases.  (path_sweep can return geometry as a module or it can return a vnf or a transform list as a function.)    I think you maybe can get a list of paths out of vnf_halfspace in addition to a vnf.  Not sure actually if there is a module.  (Maybe if there isn't there should be?)

I see no advantage of having the combined mouseover, though.

**Comment by RAMilewski** [comment]
> I see no advantage of having the combined mouseover, though.

Fewer things for the user to click on.

**Comment by revarbat** [comment]
> Personally, I favor a hover text rather than a footnote link that takes the reader away from the module or function in question.

The hover text can be done with `<sup title="hover text">TEXT</sup>` without the link that the file footers are trying.


> FWIW the STD links in the sidebar don't appear to point to actual content yet. ...at least not in Firefox or Chrome.

Oddly, even though I'm linking to the File Footnotes header in the exact same way as I link to other headers, that link just won't work.  I cannot determine why.

**Comment by RAMilewski** [comment]
My experience is that the mouseover links don't survive docsgen.   

The advantage of using links in the mouseovers is that would eliminate a huge amount of redundant typing since there would be at most a dozen or so possible mouseover texts.

**Comment by revarbat** [comment]
If you add a line like:

>// SynTags: VNF,Geom

after the Synopsis line, it will show those tags wherever the Synopsis is shown.
Current tags are defined in the .openscad_docsgen_rc:

>DefineSynTags:
>  VNF = Returns a VNF when called as a function.
>  Path = Returns a Path when called as a function.
>  Region = Returns a Region when called as a function.
>  Geom = Returns Geometry when called as a module.

**Comment by revarbat** [comment]
Docsgen v2.0.37 has SynTags support.
https://github.com/revarbat/openscad_docsgen/releases/tag/v2.0.37

**Comment by revarbat** [comment]
Oh, I've added SynTags to shapes3d.scad.

**Comment by revarbat** [comment]
Mouseovers verified to work in Safari (and therefore probably other WebKit based browsers like Chrome) and FireFox.

**Comment by adrianVmariano** [comment]
Mouseovers work OK under Firefox Windows, but the actual display of the titles is broken.  Below is the manual entry for prismoid:

![image](https://user-images.githubusercontent.com/26014420/231821751-769f0d25-3dd6-4ec4-acae-9bd9747e11d8.png)

**Comment by adrianVmariano** [comment]
I feel like the tags look funny as superscripts in this context.  I think they should just be inline at regular text height.  I also think making them links to something might be nice, so that they look "active" and users realize they can hover.   The link could be the same for all tags and go to a page listing the various BOSL2 types, e.g. what is a region, path, VNF, etc.  

Also, the text for VNF I think should say "Can return a VNF when called as a function".  We have functions that sometimes return a VNF but also can return something else.  (path_sweep does this).  I think turtle can return a path, but if you ask for the full state you get that instead.  So the statement "Returns a VNF..." isn't completely accurate for some functions that have multiple possible returns.

**Comment by RAMilewski** [comment]
We're also going to need 
 `[Mat]  = Can return a transformation matrix.`

...but as we discover more they're easy to add to opensecad_docsgen_rc.

**Comment by adrianVmariano** [comment]
Also `[MatList] = Can return a list of transformation matrices...`

**Comment by revarbat** [comment]
If all the SynTags end with "when called as a function" or "when called as a module", then are those clauses needed at all?
I'm proposing something like:
```
DefineSynTags:
  Geom = Can return geometry.
  Mat = Can return a transformation matrix.
  MatList = Can return a list of transformation matrices.
  Path = Can return a Path.
  Region = Can return a Region.
  Trans = Can transform children.
  VNF = Can return a VNF.
```

**Comment by adrianVmariano** [comment]
Seems kinda like the main reason to add remarks like "when called as a function" or "when called as a module" would be to provide clarity for new users who are maybe muddled about the module/function distinction.  I don't know if that's truly helpful or not.

**Comment by adrianVmariano** [comment]
Why does github now add a smily face to every post?

**Comment by revarbat** [comment]
It's a button to let you add thumbs-up, thumbs-down, and other simple reaction replies to a comment.

**Comment by revarbat** [comment]
Eh, I don't think the purpose of a Synopsis is to clarify things for newbies.  It's purpose is to give a hint as to the purpose of a function/module/constant in the indices.

**Comment by revarbat** [comment]
Clarity as to the specifics of a call are a part of the longer Description, Arguments or other parts of the main entry.

**Comment by revarbat** [comment]
I think the only beziers are in beziers.scad.

**Comment by RAMilewski** [comment]
Transforms can move bezier patches.   ....or are those just lists?

**Comment by revarbat** [comment]
A bezier patch is just a list of pointlists.

**Comment by RAMilewski** [comment]
assert_equal() and assert_approx() currently have the same description, but slightly different code.
I see no provision to supply a value for epsilon in assert_approx(), so how are these different?

**Comment by adrianVmariano** [comment]
There seem to be two differences.  assert_approx is calling approx().  It uses the default epsilon value, so that would be 1e-9.  assert_equal() is called != so it's testing exact equality.  It also returns true if both values are NaN, whereas assert_approx will return false in that case.   (Note, in case you don't know, NaN is  the only number that tests as not equal to itself, so special handling is needed to get nan to test as equal to itself---this behavior should be called out in the docs.)

**Comment by RAMilewski** [comment]
Ok.   Where is the default value for epsilon documented?  1e-9 is what I remembered, but I looked again and didn't find it.

approx() doesn't mention a default value for epsilon either.  I'll fix that too.

**Comment by adrianVmariano** [comment]
It's the same as all the functions near it in comparisons.scad:  it uses EPSILON which is 1e-9.

**Comment by adrianVmariano** [comment]
It looks like this is almost done.   I think only these items remain:

- [ ] coords.scad
- [ ] math.scad
- [ ] geometry.scad
- [ ] linalg.scad

There is metric_screws.scad, but that's supposed to go away.   It seems that then we need to make the full checklist for adding SynTags entries and go over stuff again.

**Comment by revarbat** [comment]
`EPSILON` is in math.scad

**Comment by adrianVmariano** [comment]
Sorta seems like SynTags should be called ReturnTypes or something.  But probably not actually worth messing with.

Are we putting tags for things like boolean, integer, number, list, and so on?

**Comment by revarbat** [comment]
In theory, SynTags can be used for things other than common return types. Consider:

    DefineSynTags:
        Slow = This can be very slow to process.
        Adv = This is an advanced level function/module, and may be difficult to use.

I floated the idea, a couple years ago, of having a Returns: block.  It can even be done without any alteration to DocsGen. It'd be treated like a Description block, just with a different header name shown. It would just be a lot of documentation to update at this point.

**Comment by RAMilewski** [comment]
+1 for Slow and Adv tags.    And perhaps SynTags should be DocTags, or just Tags?

**Comment by adrianVmariano** [comment]
Ok, so we have Synopses done, but now we need the tags.  Should we open a new issue and you can put a new checklist for that?   This one has gotten a little unwieldy.

**Comment by revarbat** [comment]
I'm working on fnliterals.scad, but otherwise, yes.

**Comment by revarbat** [comment]
fnliterals.scad is finished.

**Comment by revarbat** [comment]
For all the ones I've been doing, I've been adding Synopsis:, making sure that they have Topics:, and almost always See Also: as well.
